### PR TITLE
Use flake8 rather than pycodestyle

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes https://github.com/pydata/xarray/issues/2990
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

I also ran `isort`, which created more changes that I would have hoped. Lmk if that's OK. 
We could also enable `isort` by default. It's a balance: it reduces the chance of merge conflicts and adds automatic consistency, but at the cost of making first-time contributions a bit more difficult.

### CODE REVIEW

**Comment by max-sixty** [comment]
Also fixes https://github.com/pydata/xarray/issues/2833

**Comment by max-sixty** [comment]
Will wait for a couple of days to see if there's a response on the `pep8speaks` config issue; otherwise will revert to the existing config.

Tests pass apart from `dask-dev` which breaks on master

**Comment by max-sixty** [comment]
This is ready to merge after a review. `pep8speaks` [checks the config in `master`](https://github.com/OrkoHunter/pep8speaks/issues/130#issuecomment-500784182), and so this needs to be merged in order for the config to be updated

File: ci/requirements-py37.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ dependencies:
   - pytest-cov
   - pytest-env
   - coveralls
-  - pycodestyle
+  - flake8
```
Comment (shoyer) [inline ci/requirements-py37.yml]: Can you also replace pycodestyle with flake8 in requirements-py36.yml? Our contributing guide is still pointing users to that for setup.

**Comment by shoyer** [comment]
Thanks!

**Pull request merged**